server/src/main/java/org/opensearch/index/mapper/DerivedFieldSupportedTypes.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.mapper;
+
+import org.apache.lucene.document.DoubleField;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.FieldType;
+import org.apache.lucene.document.InetAddressPoint;
+import org.apache.lucene.document.KeywordField;
+import org.apache.lucene.document.LatLonPoint;
+import org.apache.lucene.document.LongField;
+import org.apache.lucene.document.LongPoint;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.IndexableField;
+import org.opensearch.Version;
+import org.opensearch.common.Booleans;
+import org.opensearch.common.lucene.Lucene;
+import org.opensearch.common.network.InetAddresses;
+
+import java.net.InetAddress;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Contains logic to get the FieldMapper for a given type of derived field. Also, for a given type of derived field,
+ * it is used to create an IndexableField for the provided type and object. It is useful when indexing into
+ * lucene MemoryIndex in {@link org.opensearch.index.query.DerivedFieldQuery}.
+ */
+enum DerivedFieldSupportedTypes {
```

Comment:
Not sure how doable it is, but would it be possible to use more logic form the original `FieldMapper` subclasses.

For example, the `IndexableField` that we want to create has logic implemented by `parseCreateField` in the corresponding `FieldMapper`. I'm wondering if we can plumb through the `FieldMapper` and delegate to its `parse` method, which calls `parseCreateField`.

I think if you can create a `ParseContext` instance, you could use `createExternalValueContext` to pass in the value as the appropriate type (which isn't a perfect solution, since you already need the type conversion), then `switchDoc` to set the Lucene document that you're constructing. Then a call to `mapper.parse(parseContext)` will delegate to `parseCreateField`, which will add the field(s) to the document.

Reply:
Make sense :+1: